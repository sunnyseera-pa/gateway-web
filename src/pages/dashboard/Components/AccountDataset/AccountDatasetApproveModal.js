import { Formik } from 'formik';
import React, { Suspense } from 'react';
import { Button, Form, Modal } from 'react-bootstrap';
import { useTranslation } from 'react-i18next';
import * as Yup from 'yup';
import { STATUS_INREVIEW, TEXTAREA_ROWS } from '../../../../configs/constants';
import { ReactComponent as CloseButtonSvg } from '../../../../images/close-alt.svg';
import datasetOnboardingService from '../../../../services/dataset-onboarding/dataset-onboarding';
import './AccountDatasetDecisionModal.scss';

const AccountDatasetApproveModal = ({ id, open, closed, goToNext, showGoToNext, handleApprove }) => {
	const { t } = useTranslation();

	const approveDataset = React.useCallback(
		async values => {
			const payload = {
				...values,
				id,
				applicationStatus: 'approved',
			};

			await datasetOnboardingService.putDatasetOnboarding(id, payload);
		},
		[id]
	);

	const handleSubmit = React.useCallback(
		async values => {
			await approveDataset(values);

			handleApprove({
				publisher: '',
				tab: STATUS_INREVIEW,
				message: `You have approved the dataset`,
			});
		},
		[id]
	);

	return (
		<Suspense fallback={t('loading')}>
			<Modal show={open} onHide={closed} className='decisionModal' size='lg' aria-labelledby='contained-modal-title-vcenter' centered>
				<div className='decisionModal-header'>
					<div className='decisionModal-header--wrap'>
						<div className='decisionModal-head'>
							<h1 className='black-20-semibold'>{t('dataset.approvalModal.title')}</h1>
							<CloseButtonSvg className='decisionModal-head--close' onClick={closed} />
						</div>
					</div>
				</div>
				<Formik
					initialValues={{
						applicationStatusDesc: '',
					}}
					validationSchema={Yup.object({
						applicationStatusDesc: Yup.string().max(1500, 'Description must be less than 1500 characters'),
					})}
					onSubmit={handleSubmit}>
					{({ values, errors, isValid, handleChange, handleBlur, handleSubmit }) => (
						<Form onSubmit={handleSubmit} onBlur={handleBlur}>
							<div className='decisionModal-body'>
								<div className='decisionModal-body--wrap'>
									<p data-testid='description'>{t('dataset.approvalModal.description')}</p>
									<Form.Group>
										<label for='applicationStatusDesc' className='black-14'>
											{t('dataset.approvalModal.applicationStatus')}
											<span>{values.applicationStatusDesc.length}/1500</span>
										</label>
										<Form.Control
											as='textarea'
											data-testid='dataset-reject-applicationStatusDesc'
											id='applicationStatusDesc'
											name='applicationStatusDesc'
											type='text'
											onChange={handleChange}
											value={values.applicationStatusDesc}
											onBlur={handleBlur}
											rows={TEXTAREA_ROWS}
										/>
										{errors.applicationStatusDesc ? <div className='errorMessages'>{errors.applicationStatusDesc}</div> : null}
									</Form.Group>
								</div>
							</div>
							<div className='decisionModal-footer'>
								<div data-testid='button-container' className='decisionModal-footer--wrap'>
									<Button
										className='button-secondary'
										onClick={() => {
											closed();
										}}>
										{t('dataset.approvalModal.buttons.cancel')}
									</Button>
									<Button
										disabled={!isValid}
										type='submit'
										data-testid='approve-button'
										className='button-secondary'
										style={{ marginLeft: '10px' }}>
										{t('dataset.approvalModal.buttons.approve')}
									</Button>
									<Button
										disabled={!showGoToNext || !isValid}
										className='button-secondary'
										style={{ marginLeft: '10px' }}
										onClick={async () => {
											await approveDataset(values);

											goToNext();
										}}>
										{t('dataset.approvalModal.buttons.approveAndGoToNext')}
									</Button>
								</div>
							</div>
						</Form>
					)}
				</Formik>
			</Modal>
		</Suspense>
	);
};

export default AccountDatasetApproveModal;
