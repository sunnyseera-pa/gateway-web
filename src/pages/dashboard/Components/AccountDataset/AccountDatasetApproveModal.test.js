import { render, within, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import AccountDatasetApproveModal from './AccountDatasetApproveModal';
import datasetOnboardingService from '../../../../services/dataset-onboarding/dataset-onboarding';

jest.mock('../../../../services/dataset-onboarding/dataset-onboarding');

const queryClient = new QueryClient({
	defaultOptions: {
		queries: {
			retry: false,
		},
	},
});

let containerDiv;
const goToNext = jest.fn();
const closed = jest.fn();

describe('Given the AccountDatasetApproveModal component', () => {

    const props = {
        id: "id",
        open: true,
        closed,
        goToNext,
        showGoToNext: true
    };

    describe('When it is rendered', () => {
        let wrapper;

        beforeAll(() => {
            containerDiv = createPortalContainer();
            wrapper = render(
                <QueryClientProvider client={queryClient}>
                    <AccountDatasetApproveModal {...props} container={containerDiv} />
                </QueryClientProvider>
            );
        });

        afterAll(() => {
			removePortalContainer(containerDiv);
		});

        it('Should match the snapshot', async () => {
            expect(containerDiv).toMatchSnapshot();
        });

        it('Then the Approve and go to next button should not be disabled', () => {
            const { getByText } = wrapper;
            const approveAndGoToNextButton = getByText('Approve and go to next');
            expect(approveAndGoToNextButton).not.toHaveAttribute('disabled');
        });

        describe('And showGoToNext is false', () => {
            beforeAll(() => {
                const { rerender } = wrapper;
                const newProps = {
                    ...props,
                    showGoToNext: false
                };

                rerender(
                    <QueryClientProvider client={queryClient}>
                        <AccountDatasetApproveModal {...newProps} container={containerDiv} />
                    </QueryClientProvider>
                )
            });
            it('Then the Approve and go to next button should be disabled', () => {
                const { getByText } = wrapper;
                const approveAndGoToNextButton = getByText('Approve and go to next');
                expect(approveAndGoToNextButton).toHaveAttribute('disabled');
            });
        });

        describe('And the Approve button is clicked', () => {
            let button;

            beforeAll(() => {
                const { rerender } = wrapper;

                rerender(
                    <QueryClientProvider client={queryClient}>
                        <AccountDatasetApproveModal {...props} container={containerDiv} />
                    </QueryClientProvider>
                )

                const { getByTestId } = wrapper;
                button = within(getByTestId('button-container')).getAllByText('Approve')[0];
                fireEvent.click(button);
            });

            it('Then submits the dataset approval request', async () => {
                await waitFor(() => expect(datasetOnboardingService.usePutDatasetOnboarding).toHaveBeenCalledWith({
                    id: 'id',
                    applicationStatus: 'approved',
                    applicationStatusDesc: ''
                }));
            });

            it('Then closes the modal', async () => {
                await waitFor(() => expect(closed).toHaveBeenCalled());
            });
        });

        describe('And the Approve and go to next button is clicked', () => {
            let button;

            beforeAll(() => {
                const { rerender } = wrapper;

                rerender(
                    <QueryClientProvider client={queryClient}>
                        <AccountDatasetApproveModal {...props} container={containerDiv} />
                    </QueryClientProvider>
                )
                
                const { getByText } = wrapper;
                button = getByText('Approve and go to next');
                fireEvent.click(button);
            });

            it('Then submits the dataset approval request', async () => {
                await waitFor(() => expect(datasetOnboardingService.usePutDatasetOnboarding).toHaveBeenCalledWith({
                    id: 'id',
                    applicationStatus: 'approved',
                    applicationStatusDesc: ''
                }));
            });

            it('Then goes to next dataset', async () => {
                await waitFor(() => expect(goToNext).toHaveBeenCalled());
            });

            it('Then closes the modal', async () => {
                await waitFor(() => expect(closed).toHaveBeenCalled());
            });
        });
    });
});
