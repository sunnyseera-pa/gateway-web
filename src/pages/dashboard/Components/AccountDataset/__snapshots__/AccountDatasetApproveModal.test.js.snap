// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the AccountDatasetApproveModal component When it is rendered Should match the snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;
